karapace/kafka_rest_apis/__init__.py

```diff
@@ -607,36 +608,65 @@ async def validate_publish_request_format(self, data: dict, formats: dict, conte
 
-    async def produce_message(self, *, topic: str, key: bytes, value: bytes, partition: int = None) -> dict:
-        try:
-            producer = await self._maybe_create_async_producer()
+    async def produce_messages(self, *, topic: str, prepared_records: List) -> List:
+        producer = await self._maybe_create_async_producer()
 
+        produce_futures = []
+        for key, value, partition in prepared_records:
             # Cancelling the returned future **will not** stop event from being sent, but cancelling
             # the ``send`` coroutine itself **will**.
             coroutine = producer.send(topic, key=key, value=value, partition=partition)
 
-            # Schedule the co-routine, it will be cancelled if the it is not complete in
+            # Schedule the co-routine, it will be cancelled if is not complete in
             # `self.kafka_timeout` seconds.
-            future = await asyncio.wait_for(fut=coroutine, timeout=self.kafka_timeout)
-
-            result = await future
-            return {
-                "offset": result.offset if result else -1,
-                "partition": result.topic_partition.partition if result else 0,
-            }
-        except AssertionError as e:
-            self.log.exception("Invalid data")
-            return {"error_code": 1, "error": str(e)}
-        except (KafkaTimeoutError, asyncio.TimeoutError):
-            self.log.exception("Timed out waiting for publisher")
-            # timeouts are retriable
-            return {"error_code": 1, "error": "timed out waiting to publish message"}
-        except BrokerResponseError as e:
-            self.log.exception(e)
-            resp = {"error_code": 1, "error": e.description}
-            if hasattr(e, "retriable") and e.retriable:
-                resp["error_code"] = 2
-            return resp
+            future = asyncio.wait_for(fut=coroutine, timeout=self.kafka_timeout)
+            produce_futures.append(future)
+
+        produce_results = []
+        end_results = []
+        for result in await asyncio.gather(*produce_futures, return_exceptions=True):
+            if not isinstance(result, Exception):
+                end_results.append(result)
+            elif isinstance(result, asyncio.TimeoutError):
+                self.log.exception("Timed out waiting for publisher buffer")
+                # timeouts are retriable
+                produce_results.append(
+                    {"error_code": 1, "error": "timed out waiting to publish message, producer buffer full"}
+                )
+            elif isinstance(result, AssertionError):
+                self.log.exception("Invalid data")
+                produce_results.append({"error_code": 1, "error": str(result)})
+            else:
+                self.log.exception("Unexpected exception")
+                produce_results.append({"error_code": 1, "error": str(result)})
+
+        for result in await asyncio.gather(*end_results, return_exceptions=True):
```

Comment:
Shouldn't the response results match the order of the request? Now due calling `asyncio.gather` twice there can be mismatch of the ordering of the elements if exception is one from the first set.

Reply:
I changed the first gather to create completed Future on exceptions. The results are now in order and second gather will collect the results and exceptions.